Cross compiling fixes, improvements and additions

eigendude opened the pull request:
## Description of the Change

This PR contains the build system fixes, improvements and additions broken out from https://github.com/filecoin-project/cpp-filecoin/pull/114.

### Structure of the PR

The PR is organized as follows:

1. Fixes needed for cross-compiling
2. Improvements needed for cross-compiling

### Fixes for cross-compiling

Commits:

- ~~CMake: Fix error including generated protobuf headers~~
- ~~Fix tinycbor include directory~~

These commits come first, because they address what could be considered defects. Inclusion would be valuable even if the rest of the PR is omitted. Further description is in the commit message.

### Changes for cross-compiling

Commits:

- [CMake: Expose CMake options to includes](https://github.com/eigendude/cpp-filecoin/commit/cross-compile-fixes~4)
- [CMake: Exclude test-related depends when building tests](https://github.com/eigendude/cpp-filecoin/commit/cross-compile-fixes~3)
- [CMake: Allow excluding git submodule depends](https://github.com/eigendude/cpp-filecoin/commit/cross-compile-fixes~2)

These changes enable more control over the build process when embedded in another build system.

CMake is controlled by user variables; the first one we need is `-DHUNTER_ENABLED=OFF` so that we can have control over dependencies.

Another useful variable is `-DTESTING=OFF`, because embedding test infra drastically increases the up-front effort. The first two commits above allow this variable to exclude test infra from the dependency discovery in addition to the build.

Finally, I introduced a new variable: `-DBUILD_INTERNAL_DEPS=OFF`. This causes the in-tree dependencies to be skipped. I needed this for a build system that doesn't support git submodules. Instead, the dependencies are provided by the build sytsem.

## Dependencies

- [x] https://github.com/soramitsu/libp2p/pull/59 needed for libp2p dependency.

## How has this been tested?

Temporarily disabled proofs and ran `make` from Kodi's build deps. Build succeeds.

Ran `make install`:

```
make[2]: *** No rule to make target 'install'.  Stop.
```

Next PR will address this.

igor-egorov commented on cmake/dependencies.cmake:
```diff
@@ -1,26 +1,32 @@
 
 # https://docs.hunter.sh/en/latest/packages/pkg/Microsoft.GSL.html
 hunter_add_package(Microsoft.GSL)
-find_package(Microsoft.GSL CONFIG REQUIRED)
```
According to the Hunter [docs](https://hunter.readthedocs.io/en/latest/index.html) (or [here](https://hunter.readthedocs.io/en/latest/quick-start/boost-components.html)), `find_package` is still required.

Could you please check if the code below works in your environment?
(let's keep find_package without CONFIG)
```diff
-find_package(Microsoft.GSL CONFIG REQUIRED)
+find_package(Microsoft.GSL REQUIRED)
```

eigendude replied:
I remember trying this line with no luck. CMake has no `FindMicrosoft.GSL.cmake`, but if I add an empty one, it works. The same with cppcodec - we don't use the output `CPPCODEC_FOUND` anywhere, I just added the pkg-config because I noticed cppcodec provided pkg-config files.

Does Hunter do any magic by providing found package info to CMake? If I removed all find_package() calls, I suspect fuhon would still build, does this sound weird?

igor-egorov replied:
According to [the faq](https://github.com/cpp-pm/hunter/blob/master/docs/faq/why-do-we-need-hunter-add-package.rst), `hunter_add_package(foo)` will download and install `foo` package while `find_package(foo)` should only look up for files in read-only mode.

`find_package` is a CMake builtin function, which loads package settings. Thus, `find_package` looks to be required while even when hunter is not used.

eigendude replied:
Dropped this commit

eigendude commented:
https://github.com/soramitsu/libp2p/pull/59 is declared as a dependency in the PR description. I updated the bullet to a check box so we can see that the dependency is fulfilled directly from this PR.

eigendude commented:
I updated https://github.com/soramitsu/libp2p/pull/59 and this PR. I removed all CONFIG commits. Also dropped a fix not needed after the rust BLS signature library was dropped in https://github.com/filecoin-project/cpp-filecoin/pull/129. Only 5 commits left in this PR now.

eigendude commented:
Is there more to do on the PR?

igor-egorov commented:
Hi @eigendude and sorry for the long waiting
It looks the time to merge the pr has come!
Could you please update the pr and resolve the merge conflicts?

eigendude commented:
Back from the dead! I rebased the PR and compile-tested. I dropped the tinycbor header change, it's no longer needed. I also added a commit removing the c-ares CMake declaration, as I can't find a use throughout the codebase.
